File: client/src/main/java/de/qabel/desktop/ui/actionlog/item/renderer/partial/EmojiRenderer.java
package de.qabel.desktop.ui.actionlog.item.renderer.partial;

import com.vdurmont.emoji.EmojiManager;
import com.vdurmont.emoji.EmojiParser;
import de.qabel.desktop.ui.util.Icons;
import javafx.geometry.Insets;
import javafx.scene.Node;
import javafx.scene.control.Label;
import javafx.scene.image.ImageView;
import javafx.scene.text.Text;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class EmojiRenderer extends AbstractPatternExtractionRenderer {
    private static final Pattern ALIAS_PATTERN = Pattern.compile("(:(\\w|\\||\\-)+:)");
    private static final Logger logger = LoggerFactory.getLogger(EmojiRenderer.class);

    @Override
    public boolean needsFormatting(String text) {
        return super.needsFormatting(EmojiParser.parseToAliases(text));
    }

    @Override
    protected Pattern getPattern() {
        return ALIAS_PATTERN;
    }

    @Override
    public List<Node> render(String text) {
        return super.render(EmojiParser.parseToAliases(text));
    }

    @Override
    protected Node replace(String match, Matcher matcher) {
        if (!match.startsWith(":") || !match.endsWith(":")) {
            return new Text(match);
        }
        try {
            ImageView icon = Icons.getIcon(EmojiManager.getForAlias(match), 24);
            icon.getStyleClass().add("emoji");
            Label emojiLabel = new Label("", icon);
            emojiLabel.getStyleClass().add("emoji-container");
            emojiLabel.setPadding(new Insets(0, 1, 0, 1));
            return emojiLabel;
        } catch (Exception e) {
            logger.error("failed to render emoji " + match, e);
            return renderFallback(match);
        }
    }

    @NotNull
    private Text renderFallback(String match) {
        return new Text(EmojiParser.parseToUnicode(match));
    }
}

File: client/src/main/java/de/qabel/desktop/ui/actionlog/item/renderer/partial/EmojiRenderer.kt
package de.qabel.desktop.ui.actionlog.item.renderer.partial

import com.vdurmont.emoji.EmojiManager
import com.vdurmont.emoji.EmojiParser
import de.qabel.core.logging.QabelLog
import de.qabel.desktop.ui.util.IconProvider
import javafx.geometry.Insets
import javafx.scene.Node
import javafx.scene.control.Label
import javafx.scene.text.Text
import java.util.regex.Matcher
import java.util.regex.Pattern

class EmojiRenderer(val icons: IconProvider) : AbstractPatternExtractionRenderer(), QabelLog {
    private val ALIAS_PATTERN = Pattern.compile("(:(\\w|\\||\\-)+:)")

    override fun needsFormatting(text: String) = super.needsFormatting(EmojiParser.parseToAliases(text))
    override fun getPattern() = ALIAS_PATTERN
    override fun render(text: String) = super.render(EmojiParser.parseToAliases(text))

    override fun replace(match: String, matcher: Matcher): Node {
        if (!match.startsWith(":") || !match.endsWith(":")) {
            return Text(match)
        }

        try {
            val icon = icons.getIcon(EmojiManager.getForAlias(match), 24)
            icon.styleClass.add("emoji")
            val emojiLabel = Label("", icon)
            emojiLabel.styleClass.add("emoji-container")
            emojiLabel.padding = Insets(0.0, 1.0, 0.0, 1.0)
            return emojiLabel
        } catch (e: Exception) {
            error("failed to render emoji " + match, e)
            return renderFallback(match)
        }

    }

    private fun renderFallback(match: String): Text = Text(EmojiParser.parseToUnicode(match))
}
